proxyd/integration_tests/ws_test.go

```diff
@@ -0,0 +1,253 @@
+			default:
+				_ = backendToProxyConn.WriteMessage(websocket.TextMessage, []byte("garbage"))
+			}
+		}
+	}()
+
+	// spam invalid RPCs
+	go func() {
+		for {
+			select {
+			case <-quitC:
+				return
+			default:
+				_ = client.WriteMessage(websocket.TextMessage, []byte("{\"id\": 1, \"method\": \"eth_foo\", \"params\": [\"newHeads\"]}"))
+			}
+		}
+	}()
+
+	// 1 second is enough to trigger the panic without locks in place
+	time.Sleep(time.Second)
+	close(quitC)
+}
+
+type backendHandler struct {
+	msgCB   MockWSBackendOnMessage
+	closeCB MockWSBackendOnClose
+}
+
+type clientHandler struct {
+	cb ProxydWSClientOnMessage
+}
+
+func TestWS(t *testing.T) {
+	backendHdlr := &backendHandler{
+		msgCB:   func(conn *websocket.Conn, msgType int, data []byte) {},
+		closeCB: func(conn *websocket.Conn, err error) {},
+	}
+	clientHdlr := &clientHandler{
+		cb: func(msgType int, data []byte) {},
+	}
+
+	backend := NewMockWSBackend(nil, func(conn *websocket.Conn, msgType int, data []byte) {
+		backendHdlr.msgCB(conn, msgType, data)
+	}, func(conn *websocket.Conn, err error) {
+		backendHdlr.closeCB(conn, err)
+	})
+	defer backend.Close()
+
+	require.NoError(t, os.Setenv("GOOD_BACKEND_RPC_URL", backend.URL()))
+
+	config := ReadConfig("ws")
+	shutdown, err := proxyd.Start(config)
+	require.NoError(t, err)
+	client, err := NewProxydWSClient("ws://127.0.0.1:8546", func(msgType int, data []byte) {
+		clientHdlr.cb(msgType, data)
+	}, nil)
+	defer client.HardClose()
+	require.NoError(t, err)
+	defer shutdown()
+
+	tests := []struct {
+		name       string
+		backendRes string
+		expRes     string
+		clientReq  string
+	}{
+		{
+			"ok response",
+			"{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":\"0xcd0c3e8af590364c09d0fa6a1210faf5\"}",
+			"{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":\"0xcd0c3e8af590364c09d0fa6a1210faf5\"}",
+			"{\"id\": 1, \"method\": \"eth_subscribe\", \"params\": [\"newHeads\"]}",
+		},
+		{
+			"garbage backend response",
+			"gibblegabble",
+			"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32013,\"message\":\"backend returned an invalid response\"},\"id\":null}",
+			"{\"id\": 1, \"method\": \"eth_subscribe\", \"params\": [\"newHeads\"]}",
+		},
+		{
+			"blacklisted RPC",
+			"}",
+			"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32001,\"message\":\"rpc method is not whitelisted\"},\"id\":1}",
+			"{\"id\": 1, \"method\": \"eth_whatever\", \"params\": []}",
+		},
+		{
+			"garbage client request",
+			"{}",
+			"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32700,\"message\":\"parse error\"},\"id\":null}",
+			"barf",
+		},
+		{
+			"invalid client request",
+			"{}",
+			"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32700,\"message\":\"parse error\"},\"id\":null}",
+			"{\"jsonrpc\": \"2.0\", \"method\": true}",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			timeout := time.NewTicker(30 * time.Second)
+			doneCh := make(chan struct{}, 1)
+			backendHdlr.msgCB = func(conn *websocket.Conn, msgType int, data []byte) {
```

Comment:
The msgCB is read in a separate goroutine in `MockWSBackend.ServeHTTP`. This races with the write here and would make the test flaky.

Ensure that access to all `backendHandler` callbacks are exclusive.

Reply:
Strange that the race detector didn't pick this up. Will fix.

Reply:
it's probably because the `client.WriteMessage` call on L160 and the [ReadMessage](https://github.com/ethereum-optimism/optimism/pull/2711/files/62b4e9f82af4ad5f93b679fa6341a8ebd52db25c#diff-a8386cda07790cbe1234c20956a09b1c026d20cc1b77c7f771d0b647851186a9R295) in `MockWSBackend` act as memory barriers (internally in gorilla) so the cb access is inadvertently consistent. Though I wouldn't rely on this behavior especially if we upgrade the websocket module.

Reply:
Ah, that makes sense. These are hidden behind an `atomic.Value` now, so should be thread safe.